tools/rostopic/src/rostopic/__init__.py

```diff
@@ -280,50 +280,62 @@ def _rostopic_bw(topic, window_size=-1):
-    msg_attribute = pattern[1:]
-
-    # use slice arguments if present
-    array_index_or_slice_object = None
-    index = msg_attribute.find('[')
-    if index != -1:
-        if not msg_attribute.endswith(']'):
-            sys.stderr.write("Topic name '%s' contains '[' but does not end with ']'\n" % msg_attribute)
-            return None
-        index_string = msg_attribute[index + 1:-1]
-        try:
-            array_index_or_slice_object = _get_array_index_or_slice_object(index_string)
-        except AssertionError as e:
-            sys.stderr.write("Topic name '%s' contains invalid slice argument '%s': %s\n" % (msg_attribute, index_string, str(e)))
-            return None
-        msg_attribute = msg_attribute[:index]
+    evals = []  # list of (field_name, slice_object) pairs
+    fields = [f for f in pattern.split('/') if f]
+    for f in fields:
+        if '[' in f:
+            field_name, rest = f.split('[', 1)
+            rest = rest[:-1]  # slice content, removing closing bracket
+            try:
+                array_index_or_slice_object = _get_array_index_or_slice_object(rest)
+            except AssertionError as e:
+                print("field '%s' has invalid slice argument '%s': %s"
+                      % (field_name, rest, str(e)), file=sys.stderr)
+                return None
+            evals.append((field_name, array_index_or_slice_object))
+        else:
+            evals.append((f, None))
+
+    def _msgeval(msg, evals):
+        for i, (field_name, slice_object) in enumerate(evals):
+            try: # access field first
+                msg = getattr(msg, field_name)
+            except AttributeError:
+                print("no field named %s in %s" % (field_name, pattern), file=sys.stderr)
+                return None
+
+            if slice_object is not None: # access slice
+                try:
+                    msg = msg.__getitem__(slice_object)
+                except IndexError as e:
+                    print("%s: %s" % (str(e), pattern), file=sys.stderr)
+                    return None
+
+                # if a list is returned here (i.e. not only a single element accessed),
+                # we need to recursively call msg_eval() with the rest of evals
+                # in order to handle nested slices
+                if isinstance(msg, list):
+                    rest = evals[i + 1:]
+                    return [_msgeval(m, rest) for m in msg]
+        return msg
 
     def msgeval(msg):
-        # I will probably replace this with some less beautiful but more efficient
-        try:
-            value = _get_nested_attribute(msg, msg_attribute)
-        except AttributeError as e:
-            sys.stdout.write("no field named [%s]"%pattern+"\n")
-            return None
-        if array_index_or_slice_object is not None:
-            value = value[array_index_or_slice_object]
-        return value
-    return msgeval
+        return _msgeval(msg, evals)
```

Comment:
Since `msgeval` is now basically empty you could move the implementation of `_msgeval` here and get rid of `_msgeval`.